Mathlib/Algebra/Polynomial/CoeffList.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2023 Alex Meiburg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Meiburg
+-/
+import Mathlib.Algebra.Polynomial.Degree.Definitions
+import Mathlib.Algebra.Polynomial.EraseLead
+import Mathlib.Data.List.Range
+
+/-!
+# A list of coefficients of a polynomial
+
+## Definition
+
+* `coeffList f`: a `List` of the coefficients, from leading term down to constant term.
+* `coeffList 0` is defined to be `[]`.
+
+This is useful for talking about polynomials in terms of list operations. It is "redundant" data in
+the sense that `Polynomial` is already a `Finsupp` (of its coefficients), and `Polynomial.coeff`
+turns this into a function, and these have exactly the same data as `coeffList`. The difference is
+that `coeffList` is intended for working together with list operations: getting `List.head`,
+comparing adjacent coefficients with each other, or anything that involves induction on Pollynomials
+by dropping the leading term (which is `Polynomial.eraseLead`).
+
+Note that `coeffList` _starts_ with the highest-degree terms and _ends_ with the constant term. This
+might seem backwards in the sense that `Polynomial.coeff` and `List.get!` are reversed to one
+another, but it means that induction on `List`s is the same as induction on
+`Polynomial.leadingCoeff`.
+
+The most significant theorem here is `coeffList_eraseLead`, which says that `coeffList P` can be
+written as `leadingCoeff P :: List.replicate k 0 ++ coeffList P.eraseLead`. That is, the list
+of coefficients starts with the leading coefficient, followed by some number of zeros, and then the
+coefficients of `P.eraseLead`.
+-/
+open Polynomial
+
+namespace Polynomial
+
+variable {α : Type*}
+
+section Semiring
+
+variable [Semiring α]
+
+/-- The list of coefficients starting from the leading term down to the constant term.
+
+Defining this with `P.support` keeps it computable, even if the base ring doesn't have
+`DecidableEq`.
+-/
+def coeffList (P : α[X]) : List α :=
+  (List.range P.degree.succ).reverse.map P.coeff
+
+variable {P : α[X]}
+
+variable (α) in
+@[simp]
+theorem coeffList_zero : (0 : α[X]).coeffList = [] := by
+  simp [coeffList]
+
+/-- Only the zero polynomial gives nil coeffList -/
+@[simp]
+theorem coeffList_eq_nil_iff (P : α[X]) : P.coeffList = [] ↔ P = 0 := by
+  simp [coeffList, ←bot_eq_zero', -Nat.bot_eq_zero]
```

Comment:
I think you should add another simp lemma `succ_eq_zero {a : WithBot α} : WithBot.succ a = 0 ↔ a = ⊥` in `Algebra.Order.SuccPred` (which you can create in case it doesn't exist already. Then this will become
```suggestion
  simp [coeffList]
```